lib/letters.ml

```diff
@@ -140,69 +140,72 @@ let build_email ~from ~recipients ~subject ~body =
-  if c.with_starttls then
-    let* res =
-      Sendmail_handler.run_with_starttls ~hostname ~port ~domain ~authentication
-        ~tls_authenticator ~from:from_addr ~recipients ~mail
+  try
+    let ( let* ) = Lwt.bind in
+    let authentication : Sendmail.authentication =
+      { username = c.username; password = c.password; mechanism = Sendmail.PLAIN }
+    in
+    let port =
+      match (c.port, c.with_starttls) with
+      | None, true -> 587
+      | None, false -> 465
+      | Some v, _ -> v
+    in
+    let mail = Mrmime.Mt.to_stream m in
+    let from_mailbox =
+      match Emile.of_string c.sender with
+      | Ok v -> v
+      | Error `Invalid -> failwith "Invalid sender address"
+    in
+    let from_addr =
+      match Colombe_emile.to_reverse_path from_mailbox with
+      | Ok v -> v
+      | Error (`Msg msg) -> failwith msg
     in
-    match res with
-    | Ok () -> Lwt.return (Ok ())
-    | Error err ->
+    let recipients =
+      List.map
+        (fun recipient ->
+           (match recipient with To a -> a | Cc a -> a | Bcc a -> a)
+           |> str_to_colombe_address)
+        r
+    in
+    let domain =
+      match domain_of_reverse_path from_addr with
+      | Ok v -> v
+      | Error _ -> failwith "Failed to extract domain of sender address"
+    in
+    let hostname =
+      match Domain_name.of_string c.hostname with
+      | Error _ -> failwith "Config hostname is not valid hostname"
+      | Ok hostname -> (
+          match Domain_name.host hostname with
+          | Error _ -> failwith "Config hostname is not valid hostname"
+          | Ok hostname -> hostname )
+    in
+    let* certs =
+      match Fpath.of_string c.ca_dir with
+      | Ok path -> certs_of_pem_directory ~ext:"pem" path
+      | Error _ -> failwith "Failed to open CA certificates directory"
+    in
+    let tls_authenticator = X509.Authenticator.chain_of_trust ~time:now certs in
+    if c.with_starttls then
+      let* res =
+        Sendmail_handler.run_with_starttls ~hostname ~port ~domain ~authentication
+          ~tls_authenticator ~from:from_addr ~recipients ~mail
+      in
+      match res with
+      | Ok () -> Lwt.return (Ok ())
+      | Error err ->
         Lwt.fail_with
           (Fmt.str "Sending email failed, %a" Sendmail_with_tls.pp_error err)
-  else
-    let* res =
-      Sendmail_handler.run ~hostname ~port ~domain ~authentication
-        ~tls_authenticator ~from:from_addr ~recipients ~mail
-    in
-    match res with
-    | Ok () -> Lwt.return (Ok ())
-    | Error err ->
+    else
+      let* res =
+        Sendmail_handler.run ~hostname ~port ~domain ~authentication
+          ~tls_authenticator ~from:from_addr ~recipients ~mail
+      in
+      match res with
+      | Ok () -> Lwt.return (Ok ())
+      | Error err ->
         Lwt.fail_with (Fmt.str "Sending email failed, %a" Sendmail.pp_error err)
+  with
+  | _ -> Lwt.fail_with "Unexpected internal failure"
```

Comment:
This fails with an exception, if you want to return a `Result.t` you should use `Lwt_result.fail`

Reply:
OK, I have to agree that my change was pretty badly prepared. How about now?